🐛 🚑  Bug on StaticSceneManager's cleanup process

This is code in `StaticSceneManager`

```java
sceneCache.subscribe((scene) -> SwingUtilities.invokeLater(() -> destroyScene(scene)));
```

The problem lies in this code below:

```java
sceneCache.remove(scene.getSceneName());
rootContainer.remove(scene.getSceneView());
scene.onDestroy();
```

Basically, we're calling remove(), and remove() is doing:

```java
@Override
public synchronized V remove(K key) {
        var val = super.remove(key);
        notifySubscribers(val);

        return val;
}
```

Thus, we have an indirect recursion unintentionally, which is blocked by this code:

```java
if (scene == null) {
	return;
}
```

It's stoped since `scene` would be null returned by `LRU.remove()` after a second invocation at this point.
So, now, if the scene correctly handles cleanups, we only get a `NullPointerException` since `scene.onDestroy()` has already been called.

Logging information:

```bash
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: com.github.hanzm_10.murico.swingapp.MuricoSwingApp.main()] [Source: com.github.hanzm_10.murico.swingapp.lib.io.PropertiesIO][22m[0m[37m
> Loading properties file: metadata.properties[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.io.PropertiesIO][22m[0m[37m
> Loading properties file: ApplicationConfig.properties[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene registered: auth[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene registered: home[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene registered: main[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene registered: login[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene registered: register[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.io.PropertiesIO][22m[0m[37m
> Loading properties file: styles.properties[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Navigated to scene: main[0m
[1m[94m[5/6/25, 3:51 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Navigated to scene: auth[0m
[1m[94m[5/6/25, 3:52 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene destroyed: main[0m
[1m[94m[5/6/25, 3:52 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene manager destroyed.[0m
[1m[94m[5/6/25, 3:52 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene destroyed: auth[0m
[1m[94m[5/6/25, 3:52 PM][36m [INFO] [1m[Thread: AWT-EventQueue-0] [Source: com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager][22m[0m[37m
> Scene manager destroyed.[0m
[WARNING] 
java.lang.NullPointerException: Cannot invoke "com.github.hanzm_10.murico.swingapp.lib.navigation.manager.SceneManager.destroy()" because "this.sceneManager" is null
	at com.github.hanzm_10.murico.swingapp.scenes.LoggedOutScene.onDestroy(LoggedOutScene.java:81)
	at com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager.destroyScene(StaticSceneManager.java:95)
	at com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager.lambda$1(StaticSceneManager.java:58)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
[WARNING] 
java.lang.NullPointerException: Cannot invoke "javax.swing.JButton.removeActionListener(java.awt.event.ActionListener)" because "this.loginButton" is null
	at com.github.hanzm_10.murico.swingapp.scenes.auth.MainAuthScene.onDestroy(MainAuthScene.java:120)
	at com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager.destroyScene(StaticSceneManager.java:95)
	at com.github.hanzm_10.murico.swingapp.lib.navigation.manager.impl.StaticSceneManager.lambda$1(StaticSceneManager.java:58)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
[INFO] [1m------------------------------------------------------------------------[m
[INFO] [1;31mBUILD FAILURE[m
[INFO] [1m------------------------------------------------------------------------[m
[INFO] Total time:  6.191 s
[INFO] Finished at: 2025-05-06T15:52:02+08:00
[INFO] [1m------------------------------------------------------------------------[m
[ERROR] Failed to execute goal [32morg.codehaus.mojo:exec-maven-plugin:3.1.0:java[m [1m(default-cli)[m on project [36mmurico-swingapp[m: [1;31mAn exception occurred while executing the Java class. Cannot invoke "com.github.hanzm_10.murico.swingapp.lib.navigation.manager.SceneManager.destroy()" because "this.sceneManager" is null[m -> [1m[Help 1][m
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the [1m-e[m switch.
[ERROR] Re-run Maven using the [1m-X[m switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [1m[Help 1][m http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

--- a/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/cache/ObserverLRU.java
+++ b/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/cache/ObserverLRU.java
@@ -1,4 +1,4 @@
-/** 
+/**
  *  Copyright 2025 Aaron Ragudos, Hanz Mapua, Peter Dela Cruz, Jerick Remo, Kurt Raneses, and the contributors of the project.
  *
  *  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”),
@@ -21,52 +21,78 @@ import com.github.hanzm_10.murico.swingapp.lib.observer.Subscriber;
 
 /** Useful for observing the LRU cache when it evicts an item. */
 public class ObserverLRU<K, V> extends LRU<K, V> implements Observer<V> {
-    protected List<Subscriber<V>> subscribers;
+	protected List<Subscriber<V>> subscribers;
 
-    public ObserverLRU(int capacity) {
-        super(capacity);
+	public ObserverLRU(int capacity) {
+		super(capacity);
 
-        subscribers = new ArrayList<>();
-    }
+		subscribers = new ArrayList<>();
+	}
 
-    @Override
-    public synchronized V remove(K key) {
-        var val = super.remove(key);
-        notifySubscribers(val);
+	@Override
+	public void notifySubscribers(V value) {
+		for (var subscriber : subscribers) {
+			subscriber.notify(value);
+		}
+	}
 
-        return val;
-    }
+	@Override
+	public synchronized V remove(K key) {
+		var val = super.remove(key);
 
-    @Override
-    protected V trimCache() {
-        var val = super.trimCache();
-        notifySubscribers(val);
+		if (val != null) {
+			notifySubscribers(val);
+		}
 
-        return val;
-    }
+		return val;
+	}
 
-    @Override
-    public void notifySubscribers(V value) {
-        for (var subscriber : subscribers) {
-            subscriber.notify(value);
-        }
-    }
+	/**
+	 * Used for whenever a caller, that's also a subscriber, does not need to be
+	 * updated since it's contextually aware that the item it's trying to remove is,
+	 * well... removed. Of course if {@code V} is null, then the subscribers will
+	 * also simply not be called.
+	 *
+	 * @param key
+	 * @param shouldPublishUpdate
+	 * @return
+	 */
+	public synchronized V remove(K key, boolean shouldPublishUpdate) {
+		var val = super.remove(key);
 
-    @Override
-    public void subscribe(Subscriber<V> subscriber) {
-        if (subscribers.contains(subscriber)) {
-            return;
-        }
+		if (val != null && shouldPublishUpdate) {
+			notifySubscribers(val);
+		}
 
-        subscribers.add(subscriber);
-    }
+		return val;
+	}
 
-    @Override
-    public void unsubscribe(Subscriber<V> subscriber) {
-        if (!subscribers.contains(subscriber)) {
-            return;
-        }
+	@Override
+	public void subscribe(Subscriber<V> subscriber) {
+		if (subscribers.contains(subscriber)) {
+			return;
+		}
 
-        subscribers.remove(subscriber);
-    }
+		subscribers.add(subscriber);
+	}
+
+	@Override
+	protected V trimCache() {
+		var val = super.trimCache();
+
+		if (val != null) {
+			notifySubscribers(val);
+		}
+
+		return val;
+	}
+
+	@Override
+	public void unsubscribe(Subscriber<V> subscriber) {
+		if (!subscribers.contains(subscriber)) {
+			return;
+		}
+
+		subscribers.remove(subscriber);
+	}
 }

--- a/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/manager/SceneManager.java
+++ b/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/manager/SceneManager.java
@@ -9,20 +9,20 @@ import com.github.hanzm_10.murico.swingapp.lib.navigation.guard.SceneGuard;
 import com.github.hanzm_10.murico.swingapp.lib.navigation.scene.Scene;
 
 public interface SceneManager {
-    void navigateTo(@NotNull final String sceneName);
+	void destroy();
 
-    void registerScene(@NotNull final String sceneName, @NotNull final SceneFactory sceneFactory);
+	String getCurrentSceneName();
 
-    void registerScene(@NotNull final String sceneName, @NotNull final SceneFactory sceneFactory,
-            @NotNull final SceneGuard sceneGuard);
+	JPanel getRootContainer();
 
-    void unregisterScene(@NotNull final String sceneName);
+	Scene getScene(@NotNull final String sceneName);
 
-    void destroy();
+	void navigateTo(@NotNull final String sceneName);
 
-    Scene getScene(@NotNull final String sceneName);
+	void registerScene(@NotNull final String sceneName, @NotNull final SceneFactory sceneFactory);
 
-    String getCurrentSceneName();
+	void registerScene(@NotNull final String sceneName, @NotNull final SceneFactory sceneFactory,
+			@NotNull final SceneGuard sceneGuard);
 
-    JPanel getRootContainer();
+	void unregisterScene(@NotNull final String sceneName);
 }

--- a/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/manager/impl/StaticSceneManager.java
+++ b/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/manager/impl/StaticSceneManager.java
@@ -50,8 +50,6 @@ public class StaticSceneManager implements SceneManager {
 		super();
 
 		cardLayout = new CardLayout();
-		cardLayout.setHgap(0);
-		cardLayout.setVgap(0);
 		rootContainer = new JPanel(cardLayout);
 		registeredSceneEntries = new HashMap<>();
 		sceneCache = new ObserverLRU<>(10);

--- a/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/manager/impl/StaticSceneManager.java
+++ b/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/manager/impl/StaticSceneManager.java
@@ -90,7 +90,7 @@ public class StaticSceneManager implements SceneManager {
 			LOGGER.severe("Scene view is null before destroy of scene: " + scene.getSceneName());
 		}
 
-		sceneCache.remove(scene.getSceneName());
+		sceneCache.remove(scene.getSceneName(), false);
 		rootContainer.remove(scene.getSceneView());
 		scene.onDestroy();
 

--- a/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/scene/Scene.java
+++ b/murico-swingapp/src/main/java/com/github/hanzm_10/murico/swingapp/lib/navigation/scene/Scene.java
@@ -2,46 +2,59 @@ package com.github.hanzm_10.murico.swingapp.lib.navigation.scene;
 
 import javax.swing.JPanel;
 
+import com.github.hanzm_10.murico.swingapp.lib.navigation.manager.SceneManager;
+
 public interface Scene {
-    String getSceneName();
+	default boolean canHide() {
+		return true;
+	}
 
-    JPanel getSceneView();
+	default boolean canShow() {
+		return true;
+	}
 
-    default boolean supportsSubScenes() {
-        return this instanceof SubSceneSupport;
-    }
+	/**
+	 *
+	 * @return The name that {@link SceneManager} will use for accessing this scene
+	 *         in its cache for navigation.
+	 */
+	String getSceneName();
 
-    default boolean canShow() {
-        return true;
-    }
+	/**
+	 * This is where a {@link Scene} should typically create its view if it doesn't
+	 * exist yet.
+	 *
+	 * @return
+	 */
+	JPanel getSceneView();
 
-    default void onBeforeShow() {
-    }
+	default Scene getSelf() {
+		return this;
+	}
 
-    default void onShow() {
-    }
+	default void onBeforeHide() {
+	}
 
-    default boolean canHide() {
-        return true;
-    }
+	default void onBeforeShow() {
+	}
 
-    default void onCannotHide() {
-    }
+	default void onCannotHide() {
+	}
 
-    default void onBeforeHide() {
-    }
+	void onCreate();
 
-    default void onHide() {
-    }
+	default boolean onDestroy() {
+		return true;
+	}
 
-    void onCreate();
+	default void onHide() {
+	}
 
-    default boolean onDestroy() {
-        return true;
-    }
+	default void onShow() {
+	}
 
-    default Scene getSelf() {
-        return this;
-    }
+	default boolean supportsSubScenes() {
+		return this instanceof SubSceneSupport;
+	}
 
 }

--- a/murico-swingapp/src/test/java/com/github/hanzm_10/murico/swingapp/lib/cache/TestsObserverLRU.java
+++ b/murico-swingapp/src/test/java/com/github/hanzm_10/murico/swingapp/lib/cache/TestsObserverLRU.java
@@ -1,0 +1,59 @@
+package com.github.hanzm_10.murico.swingapp.lib.cache;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import com.github.hanzm_10.murico.swingapp.lib.observer.Subscriber;
+
+public class TestsObserverLRU {
+	@Test
+	@DisplayName("test trimCache() to only notify subscribers when exceeding limit.")
+	void testObserver() {
+		ObserverLRU<String, String> lruObs = new ObserverLRU<String, String>(2);
+		var isCalled = new AtomicBoolean(false);
+		Subscriber<String> l = (_) -> {
+			isCalled.set(true);
+		};
+
+		lruObs.subscribe(l);
+		lruObs.update("foo", "bar");
+		lruObs.update("deez", "nuts");
+
+		assertFalse(isCalled.get());
+
+		lruObs.update("say", "what");
+
+		assertTrue(isCalled.get());
+
+		lruObs.unsubscribe(l);
+	}
+
+	@Test
+	@DisplayName("test ObserverLRU to not cause indirect recursion when remove() is called by a subscriber")
+	void testObserverIndirectRecursion() {
+		ObserverLRU<String, String> lruObs = new ObserverLRU<String, String>(2);
+		var calls = new AtomicInteger(0);
+		Subscriber<String> l2 = (s) -> {
+			// basically, if the remove() invocation inside this lambda function calls this
+			// subscriber lambda function, then calls != 0 by then. Verifies that the
+			// subscriber is only ever called once for the same key that's being removed
+			assertEquals(0, calls.getAndIncrement());
+			assertNotNull(s);
+			lruObs.remove("foo");
+		};
+
+		lruObs.update("foo", "bar");
+		lruObs.update("deez", "nuts");
+		lruObs.subscribe(l2);
+		lruObs.remove("foo");
+		lruObs.unsubscribe(l2);
+	}
+}

